File: judges/codeforces/gym/nacional-2023/g.py
def intersect(retaA, retaB):
	typeA, a, b, lxa, rxa = retaA
	typeB, c, d, lxb, rxb = retaB

	lxc, rxc = max(lxa, lxb), min(rxa, rxb)
	if lxc > rxc:
		return None
	
	if typeA == typeB == "reta":
		if a != c and b != d:
			x = (d-b)//(a-c)
			y = x * a + b
			if x >= lxc and x <= rxc:
				return ("ponto", x, y, x, x)
		elif a == c and b == d:
			return ("reta", a, b, lxc, rxc)
	elif typeA == typeB == "ponto" and a == c and b == d:
		return retaA
	elif typeA == "ponto" and typeB == "reta" and a*c+d == b:
		return ("ponto", a, b, a, a) 
	elif typeA == "reta" and typeB == "ponto" and c*a+b == d:
		return ("ponto", c, d, c, c) 
	return None

def construirRetas(a, b, c):
	retas = []
	retas.append(("reta",  1, b + c - a, a - c,		a))
	retas.append(("reta", -1, a + b + c, 	a,	a + c))
	retas.append(("reta",  1, b - a - c, 	a,	a + c))
	retas.append(("reta", -1, a + b - c, a - c, 	a))
	return retas

def construirPontos(reta):
	typeA, a, b, lxa, rxa = reta
	if typeA == "ponto":
		return [(a, b)]
	elif typeA == "reta":
		pontos = []
		for x in range(lxa, rxa+1):
			y = x * a + b
			pontos.append((x, y))
		return pontos
	else:
		return []
	

n = int(input())
a, b, c = map(int, input().split())
retas = construirRetas(a, b, c)
for i in range(n-1):
	a, b, c = map(int, input().split())
	novasRetas = construirRetas(a, b, c)
	interRetas = []

	for i in retas:
		for j in novasRetas:
			inter = intersect(i, j)
			if inter:
				interRetas.append(inter)
	retas = interRetas

pontos = []
for r in retas:
	pontos.extend(construirPontos(r))
pontos = sorted(set(pontos))
for p in pontos:
	x, y = p
	print(x, y)

File: judges/codeforces/gym/nacional-2023/g.cpp
#include<bits/stdc++.h>
#define int long long

using namespace std;

struct GEO {
	string type; // "reta", "ponto"
	int a, b, l, r; // for "reta": ax+b, and valid x in [l, r]. For "ponto": a = x, b = y.

	bool operator<(const GEO &other) const {
		if (type != other.type) {
			return type == "ponto";
		} else if (a != other.a) {
			return a < other.a;
		} else if (b != other.b) {
			return b < other.b;
		} else if (l != other.l) {
			return l < other.l;
		} else {
			return r < other.r;
		}
	}
};

GEO intersect(const GEO &retaA, const GEO &retaB) {
	string typeA = retaA.type; string typeB = retaB.type;
	int a = retaA.a, b = retaA.b, lxa = retaA.l, rxa = retaA.r;
	int c = retaB.a, d = retaB.b, lxb = retaB.l, rxb = retaB.r;
	
	int lxc = max(lxa, lxb);
	int rxc = min(rxa, rxb);
	
	if(lxc > rxc) return {};
	
	if(typeA == "reta" && typeB == "reta") {
		if(a != c) {
			double denominator = a - c;
			double x = (d - b) / denominator;
			double y = x * a + b;
			
			if(x == (int)x && y == (int)y && x >= lxc && x <= rxc) {
				return GEO{"ponto", (int)x, (int)y, (int)x, (int)x};
			}
		} else if(a == c && b == d) {
			return GEO{"reta", a, b, lxc, rxc};
		}
	} else if(typeA == "ponto" && typeB == "ponto" && a == c && b == d) {
		return retaA;
	} else if(typeA == "ponto" && typeB == "reta" && a * c + d == b) {
		return GEO{"ponto", a, b, a, a};
	} else if(typeA == "reta" && typeB == "ponto" && c * a + b == d) {
		return GEO{"ponto", c, d, c, c};
	}

	return {};
}

set<GEO> construirRetas(int a, int b, int c) {
	set<GEO> retas;
	retas.insert(GEO{"reta",  1, b + c - a, a - c,		a});
	retas.insert(GEO{"reta", -1, a + b + c,   a,    a + c});
	retas.insert(GEO{"reta",  1, b - a - c,   a,    a + c});
	retas.insert(GEO{"reta", -1, a + b - c, a - c, 		a});
	return retas;
}

vector<pair<int,int>> construirPontos(const GEO &reta) {
	vector<pair<int,int>> pontos;
	if(reta.type == "ponto") {
		pontos.emplace_back(reta.a, reta.b);
	} else if(reta.type == "reta") {
		for (int x = reta.l; x <= reta.r; x++) {
			int y = x * reta.a + reta.b;
			pontos.emplace_back(x, y);
		}
	}
	return pontos;
}

signed main(){
	ios::sync_with_stdio(false);
	cin.tie(nullptr);
	
	int n; cin>>n;
	int a, b, c; cin>>a>>b>>c;
	set<GEO> retas = construirRetas(a, b, c);
	
	for (int k = 1; k < n; k++){
		cin>>a>>b>>c;
		set<GEO> novasRetas = construirRetas(a, b, c);
		set<GEO> interRetas;
		for (const auto &r1 : retas) {
			for (const auto &r2 : novasRetas) {
				auto inter = intersect(r1, r2);
				if(inter.type != ""){
					interRetas.insert(inter);
				}
			}
		}
		retas.clear();
		retas.insert(interRetas.begin(), interRetas.end());
	}
	
	set<pair<int,int>> pontos;
	for (const auto &r : retas) {
		vector<pair<int,int>> pts = construirPontos(r);
		for(auto &p: pts) {
			pontos.insert(p);
		}
	}
	for(auto &p : pontos) {
		cout<<p.first<<" "<<p.second<<"\n";
	}
//lkasjdflçkasjlçfkdazjçklsd
	
	return 0;
}